File: backend/services/anomaly.py
from __future__ import annotations
from database import get_supabase


def compute_daily_anomaly_insights(date: str) -> dict:
    """
    supabase_schema.sql: requests'te created_at yok; orders/inventory yok.
    Zaman serisi anomalileri hesaplanamaz — düşük stok + anlık gecikme oranı.
    """
    sb = get_supabase()

    try:
        products_res = sb.table("products").select("ad, mevcut_kg, kapasite_kg").execute()
        products = products_res.data or []
    except Exception:
        products = []

    stock_depletion_anomalies = []
    for p in products:
        current = p.get("mevcut_kg") or 0
        capacity = p.get("kapasite_kg") or 1
        pct = round((current / capacity) * 100) if capacity > 0 else 100
        if pct > 30:
            continue
        stock_depletion_anomalies.append({
            "product_name": p.get("ad", ""),
            "previous_7d_consumption": 0,
            "recent_7d_consumption": current,
            "acceleration_ratio": 2.0 if pct <= 15 else 1.5,
            "is_anomaly": True,
        })

    try:
        reqs_res = sb.table("requests").select("durum").execute()
        reqs = reqs_res.data or []
    except Exception:
        reqs = []

    delayed = sum(1 for r in reqs if r.get("durum") in ("gecikti", "gecikmiş"))
    delayed_ratio = (delayed / len(reqs)) if reqs else 0

    return {
        "date": date,
        "demand_anomalies": [],
        "delayed_spike": {
            "date": date,
            "delayed_ratio_today": delayed_ratio,
            "delayed_ratio_7d_average": 0,
            "multiplier": 2 if delayed_ratio > 0.2 else 0,
            "is_anomaly": len(reqs) >= 2 and delayed_ratio >= 0.2,
        },
        "stock_depletion_anomalies": stock_depletion_anomalies,
        "producer_silence": [],
    }

File: src/lib/ai/anomaly.ts
import { createServerClient } from "@/lib/supabase/client";
import type {
  DailyAnomalyInsights,
  DelayedSpike,
  DemandAnomaly,
  ProducerSilence,
  StockDepletionAnomaly,
} from "@/lib/ai/types";

/**
 * supabase_schema.sql: requests'te created_at yok; orders/inventory/profiles yok.
 * Zaman serisi anomalileri hesaplanamaz — Python services/anomaly.py ile aynı yaklaşım:
 * stok düşüklüğü sinyali + anlık gecikme oranı.
 */
export async function computeDailyAnomalyInsights(date: string): Promise<DailyAnomalyInsights> {
  const supabase = createServerClient();

  const { data: products } = await supabase
    .from("products")
    .select("ad, mevcut_kg, kapasite_kg");

  const stockDepletionAnomalies: StockDepletionAnomaly[] = (products ?? [])
    .map((p) => {
      const current = p.mevcut_kg ?? 0;
      const capacity = p.kapasite_kg ?? 1;
      const pct = capacity > 0 ? Math.round((current / capacity) * 100) : 100;
      const is_anomaly = pct <= 30;
      return {
        product_name: p.ad ?? "",
        previous_7d_consumption: 0,
        recent_7d_consumption: current,
        acceleration_ratio: pct <= 15 ? 2 : pct <= 30 ? 1.5 : 1,
        is_anomaly,
      };
    })
    .filter((x) => x.is_anomaly)
    .sort((a, b) => b.acceleration_ratio - a.acceleration_ratio);

  const { data: reqs } = await supabase.from("requests").select("durum");
  const rows = reqs ?? [];
  const delayed = rows.filter((r) => r.durum === "gecikti" || r.durum === "gecikmiş").length;
  const delayed_ratio_today = rows.length > 0 ? delayed / rows.length : 0;

  const delayedSpike: DelayedSpike = {
    date,
    delayed_ratio_today,
    delayed_ratio_7d_average: 0,
    multiplier: delayed_ratio_today > 0.2 ? 2 : 0,
    is_anomaly: rows.length >= 2 && delayed_ratio_today >= 0.2,
  };

  return {
    date,
    demand_anomalies: [] as DemandAnomaly[],
    delayed_spike: delayedSpike,
    stock_depletion_anomalies: stockDepletionAnomalies,
    producer_silence: [] as ProducerSilence[],
  };
}
